scripts/add-erc20-tokens-coingecko.ts

```diff
@@ -0,0 +1,211 @@
+    const { id, image } = token
+    if (!id) {
+      console.warn(`⚠️ No id found for token ${token}`)
+      continue
+    }
+
+    console.log(
+      `(${i + 1}/${coingeckoResponse.data.length}) Processing token ${id}...`,
+    )
+
+    if (existingLowerCaseTokenSymbols.has(token.symbol.toLowerCase())) {
+      console.log(`Token ${id} already exists`)
+      continue
+    }
+
+    // avoid rate limit 10-30 requests / minute
+    // https://apiguide.coingecko.com/getting-started/error-and-rate-limit#rate-limit
+    await new Promise((resolve) => setTimeout(resolve, 10000))
+
+    if (!image) {
+      console.warn(`⚠️ No id or image found for token ${token}`)
+      fetchFailedTokenIds.push(id)
+      continue
+    }
+
+    // get token address from coingecko /coins/{id} endpoint, annoyingly this is
+    // not returned in the /coins/markets response
+    let address = undefined
+    let decimals = undefined
+    try {
+      console.log('Fetching token details from CoinGecko...')
+      const coinDetailResponse = await axios.get(
+        `https://api.coingecko.com/api/v3/coins/${id}`,
+        {
+          params: {
+            localization: false,
+            tickers: false,
+            market_data: false,
+            community_data: false,
+            developer_data: false,
+            sparkline: false,
+          },
+        },
+      )
+      decimals =
+        coinDetailResponse.data.detail_platforms[platformId]?.decimal_place
+      address =
+        coinDetailResponse.data.detail_platforms[platformId]?.contract_address
+      if (!address) {
+        throw new Error(`No token address found for token ${id}`)
+      }
+    } catch (error) {
+      console.warn(
+        `⚠️ Encountered error fetching token address for ${id} from Coingecko: ${error}`,
+      )
+      fetchFailedTokenIds.push(id)
+      continue
+    }
+
+    // get token metadata from token contract directly. Coingecko manually adds
+    // some info to their token list and the token symbol in particular is
+    // always (incorrectly) lowercased.
+    console.log(`Fetching token details from the contract ${address}...`)
+    const [symbol, name] = await Promise.all([
+      client.readContract({
+        address,
+        abi: erc20Abi,
+        functionName: 'symbol',
+      }),
+      client.readContract({
+        address,
+        abi: erc20Abi,
+        functionName: 'name',
+      }),
+    ])
+
+    // read the token image from coingecko and resize before saving. continue if
+    // the image cannot be saved successfully, we don't want imageless tokens.
+    try {
+      console.log('Fetching token image from Coingecko...')
+      const response = await axios.get(image, {
+        responseType: 'arraybuffer',
+        responseEncoding: 'binary',
+      })
+
+      console.log('Resizing image...')
+      const imageFile = await Jimp.read(response.data)
+      const resizedImage = await imageFile
+        .resize(256, 256)
+        .quality(100)
```

Comment:
Nothing to do here, but this reminds me it would be nice to have an automated check that ensures the images we have in the repo are optimized (with max compression, for instance with [ImageOptim](https://imageoptim.com/) or similar).